File: Exámenes/Finales/2012 FEB (DONE)/Control2Feb2012/Control2Feb2012/control2Haskell/BiPartite.hs
----------------------------------------------
-- Estructuras de Datos.  
-- 2º Curso del Grado en Ingeniería [Informática | del Software | de Computadores].
-- Escuela Técnica Superior de Ingeniería en Informática. UMA
--
-- Examen 
-- Estudio de grafos bipartitos por coloreado con búsqueda en profundidad
-- 
-- ALUMNO : Ángel Manuel Soria Gil  
-- GRADO:  Doble Grado Matemáticas + Ingeniería Informática.
--  
----------------------------------------------

module BiPartite(
    biColored    -- :: Ord v  => Graph v -> Maybe (D.Dictionary v Color)
  , Color(..)
  )  where

import Graph 
import Data.Maybe(isJust)
import qualified Dictionary as D
import qualified Stack as S


data Color = Red |  Blue deriving (Eq,Show,Ord)
nextColor Red  = Blue
nextColor Blue = Red

pushAll :: S.Stack a -> [a] -> S.Stack a
pushAll = foldr S.push   

biColored :: Ord v => Graph v -> Maybe (D.Dictionary v Color)
biColored g 
 | null vs   = Just D.empty -- empty graph is bipartite
 | otherwise = aux g D.empty (S.push (src ,Red) S.empty) 
 where 
  vs  = vertices g
  src = head vs -- initial vertex
 
aux :: Ord v => Graph v -> D.Dictionary v Color -> S.Stack (v, Color) -> 
                Maybe (D.Dictionary v Color)
aux g dict stack
   | S.isEmpty stack       = Just dict 
   | not (colored v)       = aux g (dict') (pushAll stack' listasucs)
   | c == ca               = aux g dict stack'
   | otherwise             = Nothing
-- 
-- ¡¡¡ completad el resto de guardas !! 
--
  where
     colored v          = D.isDefinedAt v dict
     stack'             = S.pop stack
     (v,c)              = S.top stack
     dict'              = D.insert v c dict
     ca                 = D.valueOf' v dict 
     listasucs          = [(vertex, nextColor c) | vertex <- (successors g v), not(colored vertex)]

--
-- ¡¡¡ completad las variables locales necesarias !!!
--
 

---------------------
--- EXAMPLES --------
---------------------
data MiVertice = A|B|C|D|E|F|G deriving (Show,Eq,Enum,Ord)

{- 
  A--B--D--F    
   \ |  |
     C--E--G
-}
g1  = mkGraphSuc vertices suc 
  where 
        vertices = [A .. G]
        suc A = [C,B]
        suc B = [A,C,D]
        suc C = [B,E]
        suc D = [B,F,E]
        suc E = [C,D,G]
        suc F = [D,D]
	suc G = [E]
-- *BiPartite> biColored g1
-- Nothing

{- 
  A--B--D--F    
     |  |
     C--E
-}
g2  = mkGraphEdges vertices edges 
  where 
        vertices =  [A .. F]
        edges = [(A,B),(B,C),(B,D),(D,E),(D,F),(C,E)]
-- *BiPartite> biColored g2
-- Just Dictionary(A->Red,B->Blue,C->Red,D->Red,E->Blue,F->Blue)

-- construcción de los bipartitos K n m
k n m = mkGraphEdges vertices edges
  where 
            vertices = [1 .. n + m]
            edges = [ (r,a) | r<-[1..n], a<-[n+1..n+m] ] 

-- *BiPartite> biColored (k 2 3)
-- Just Dictionary(1->Red,2->Red,3->Blue,4->Blue,5->Blue)

File: Exámenes/2012 FEB/prBipartite/src/graph/BiPartite.java
/**
 * @author Blas Ruiz, Data Structures, Grado en Inform‡tica. UMA.
 * Alumno: Ángel Manuel Soria Gil
 * Control 2 del 13-Febrero-2012
 * Estudio de grafos bipartitos por coloreado con búsqueda en profundidad
 */

package graph;

import dictionary.Dictionary;
import dictionary.HashDictionary;
import stack.Stack;
import stack.StackList;


public class BiPartite<V> {
	
	public static enum Color {Red, Blue;
	}

	private static Color nextColor(Color c) {
		return (c == Color.Blue) ?Color.Red:Color.Blue; 
	}
	
	private Stack<Pair<V,Color>> stack; // stack with pair of vertex and color
	private Dictionary<V,Color> dict; // dictionary: Vertices -> Color
	private boolean isBiColored;

	public BiPartite(Graph<V> graph) {
		dict = new HashDictionary<V, Color>();
		stack = new StackList<Pair<V, Color>>();
		isBiColored = true;
		if (graph.numVertices() == 0){
			return;
		}

		V src = graph.vertices().iterator().next(); // initial vertex
		stack.push(new Pair<V,Color>(src,Color.Red));
		
		while (!stack.isEmpty()) {
			Pair<V,Color> vColor = stack.top();
			stack.pop();
         	V v = vColor.first();
			Color c = vColor.second();
			if (!isDefinedAt(v, dict)){
				dict.insert(v, c);
				Color nc = nextColor(c);
				for (V vertex : graph.successors(v)){
					if (!isDefinedAt(vertex, dict)){
						Pair<V, Color> p = new Pair(vertex, nc);
						stack.push(p);
					}
				}
			} else {
				Color ac = dict.valueOf(v);
				if (!c.equals(ac)){
					isBiColored = false;
				}
			}
		} 
	}

	private boolean isDefinedAt(V v, Dictionary<V, Color> dict) {
		boolean res = false;
		for (V vertex : dict.keys()){
			if (vertex.equals(v)){
				res = true;
			}
		}
		return res;
	}


	public Dictionary<V,Color> biColored() {
		return dict;
	}
	
	public boolean isBicolored() {
		return isBiColored;
	}
	
}
